### PR TITLE
Pr unit2

### PR DESCRIPTION
**Pull request by mottaquikarim**
Rewrite of Unit 2 using Unit 1 updates as template.

LOTs to discuss here, main points:
- we should teach gulp
- promises belong in Ajax unit
- if we teach react we need to do at least a light intro to es6
- do we really have to teach d3?

### CODE REVIEW

File: unit2-apis-and-advanced-javascript/README.md [hunk 1]
```diff
@@ -1,30 +1,114 @@
...
+- Scope: .5 day
+ - Hoisting 
+ - Investigating the scope chain
+ - **Learning outcome**: Be able to answer common "what does this code output" tricky JS questions involving hoisting, scope. Examples can be seen here: https://www.sitepoint.com/5-typical-javascript-interview-exercises/ 
+- Closures: .5 day
+ - What are closures?
+ - How can they be used and when are they useful?
+ - **Learning outcome**: Solve tricky problems involving closures, examples:
+
+```js
```
Comment (bensussman) [inline unit2-apis-and-advanced-javascript/README.md:34]: I think this is too much detail to this file. Instead put this in a different (assessment, assignment, whatever) and link to it from here.

File: unit2-apis-and-advanced-javascript/README.md [hunk 2]
```diff
@@ -1,30 +1,114 @@
...
- - Firebase (or some other competitor) configuration and consumption
- - **Learning outcome**: Update the tic-tac-toe implementation from Unit 1 to store game state in Firebase, maintain leaderboards, have user accounts that users can log in to (insecurely) from any client
- - **Learning outcome**: Consume an existing API (http://jsonplaceholder.typicode.com/ or http://pokeapi.co/) and allow for the client to control different queries on the API which display different results
-4. Client Side UX Frameworks
- - React.js
- - Redux
- - d3.js
- - **Learning outcome**: Given a d3.js component (we could build it for them), design and implement a react.js app that has many d3 graphs whose state is managed by Redux
+- Building frontend javascript with Gulp: 2 days
+ - What is Gulp?
```
Comment (bensussman) [inline unit2-apis-and-advanced-javascript/README.md:12]: never heard of Gulp. Just looked it up, and read a lot of people talking about how much better than Grunt it is. 👌 

I think for small projects, it's not necessary, however we are preparing them for bigger projects in the coming months so I'm down with teaching this.

Does Gulp work for server side as well? Specifically: Node.js, Express, PostgreSQL, MongoDB?

**Comment by mottaquikarim** [comment]
@bensussman: people are opinionated! Gulp is "newer" than Grunt and has less configuration and gives the developer more control of the build tasks. Here are the slides (from a _while_ ago) that sold me to use gulp vs grunt: [slides](http://markdalgleish.github.io/presentation-build-wars-gulp-vs-grunt/)

EDIT: heh, misread your original comment! You can ignore the following links if you'd like, keeping for transparency. PS: you should also know about [webpack...](https://webpack.github.io/)
- [useful grunt things](http://markdalgleish.github.io/presentation-build-wars-gulp-vs-grunt/#5)
- [useful gulp things](http://markdalgleish.github.io/presentation-build-wars-gulp-vs-grunt/#18)

That being said, I do think it is absolutely pivotal that we teach folks about building their javascript code -- any dev shop worth it's salt is using some sort of build tool to modularize front-end javascript. (In other words, glad you're down for teaching this! It's super important)

Re: Gulp on the server side, the main use cases I have for gulp are: 
1. concatenating individual javascript modules
2. using something like [browserify](http://browserify.org/)
3. transforming jsx, etc from react
4. uglifying / etc for CSS files, JS files in production
5. transpiling to ecma5 from ecma2015

Most of the above use cases don't really exist on the server (except maybe #5) but because gulp/grunt/etc are just build tools we can hook them up to do whatever we want

**Comment by bensussman** [comment]
I have some changes i'd like to make to this, however this is definitely the right direction.

What i'm gonna do is merge this and then make my own PR making changes on top of this. 👍

**Comment by mottaquikarim** [comment]
Great -- let me know if you'd still like for my to pull out the exercises into it's own folder and link.

**Comment by bensussman** [comment]
I do, but i'll do that right now.

**Comment by mottaquikarim** [comment]
Ok thanks

**Comment by bensussman** [comment]
@mottaquikarim please update your editor to use TWO SPACES for tabs. Consistency is key!

**Comment by mottaquikarim** [comment]
@bensussman I used github's readme editor UI to make these changes -- I incorrectly assumed the spacing was set by default.

**Pull request merged**